### ISSUE TITLE
RT-DETR can not val or predict correctly.

### ISSUE THREAD

**Issue by r1cheu** [ultralytics/ultralytics#4504]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Validation, Detection

### Bug

I'm training RT-DETR on a custom dataset which works fine when using YOLOv8. RT-DETR-l also train and val normally when using yolo detect train cmd. However after finish training, when using yolo val cmd or predict with best.pt, it seems can not assign data classes correctly and detect nothing. 
```shell
# train
      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size
       5/20      3.51G     0.6381     0.5369     0.1656        128        640: 100%|██████████| 96/96 [01:00<00:00,  1.58it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 41/41 [00:02<00:00, 18.82it/s]
                   all        161       3847      0.792      0.749      0.822       0.54
# val
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 11/11 [00:17<00:00,  1.61s/it]
                   all        161       4029          0          0          0          0
```

```python

yolo detect predict model=./runs/detect/train4/weights/best.pt source=images/20230322.jpg' conf=0.3 iou=0.5
rt-detr-l summary: 498 layers, 31985795 parameters, 0 gradients

WARNING ⚠️ 'Boxes.boxes' is deprecated. Use 'Boxes.data' instead.
Traceback (most recent call last):
  File "/data/home/rlchen/app/miniconda3/envs/yolo/bin/yolo", line 8, in <module>
    sys.exit(entrypoint())
  File "/data/home/rlchen/app/miniconda3/envs/yolo/lib/python3.9/site-packages/ultralytics/cfg/__init__.py", line 431, in entrypoint
    getattr(model, mode)(**overrides)  # default args from model
  File "/data/home/rlchen/app/miniconda3/envs/yolo/lib/python3.9/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/data/home/rlchen/app/miniconda3/envs/yolo/lib/python3.9/site-packages/ultralytics/engine/model.py", line 246, in predict
    return self.predictor.predict_cli(source=source) if is_cli else self.predictor(source=source, stream=stream)
  File "/data/home/rlchen/app/miniconda3/envs/yolo/lib/python3.9/site-packages/ultralytics/engine/predictor.py", line 204, in predict_cli
    for _ in gen:  # running CLI inference without accumulating any outputs (do not modify)
  File "/data/home/rlchen/app/miniconda3/envs/yolo/lib/python3.9/site-packages/torch/utils/_contextlib.py", line 35, in generator_context
    response = gen.send(None)
  File "/data/home/rlchen/app/miniconda3/envs/yolo/lib/python3.9/site-packages/ultralytics/engine/predictor.py", line 274, in stream_inference
    s += self.write_results(i, self.results, (p, im, im0))
  File "/data/home/rlchen/app/miniconda3/envs/yolo/lib/python3.9/site-packages/ultralytics/engine/predictor.py", line 169, in write_results
    log_string += result.verbose()
  File "/data/home/rlchen/app/miniconda3/envs/yolo/lib/python3.9/site-packages/ultralytics/engine/results.py", line 287, in verbose
    log_string += f"{n} {self.names[int(c)]}{'s' * (n > 1)}, "
KeyError: 1
```

```yaml
# data.yaml
path: ./panicle_full/  # dataset root dir
train: train.txt 
val: val.txt  
test: test.txt  
names:
  0: panicle
```
![val_batch1_pred](https://github.com/ultralytics/ultralytics/assets/124009368/63c4e3cc-edba-49b6-aa89-1e2624152876)
![val_batch1_labels](https://github.com/ultralytics/ultralytics/assets/124009368/89fbf121-1f82-41bf-b85e-276e3d94a989)

### Environment

Ultralytics YOLOv8.0.158 🚀 Python-3.9.17 torch-2.0.1 CUDA:0 (NVIDIA A40, 45494MiB)

### Additional

the Ultralytics YOLOv8.0.159 or Ultralytics YOLOv8.0.160 wiill generate file directly under workdir but not runs/detect/train. 

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@r1cheu hi there,

From your description, it seems like the RT-DETR model isn't assigning the correct classes to your data during validation and prediction. This could be due to a mismatch between the classes used in training and validation/prediction. It is common for such issues to occur when the classes in the training data and the classes in the validation/prediction data are not the same.

About the error "KeyError: 1", it suggests that there might be a misalignment of the class indexing. Your yaml file shows only one class "panicle" which is assigned an index of 0. If there are predictions for class index 1, it might throw a KeyError as there's no class assigned to that index.

As for the issue of generated files directly under the workdir in Ultralytics YOLOv8.0.159 and Ultralytics YOLOv8.0.160 versions, it seems like it might be a change in how the output directory is handled in the newer versions.

I hope this helps clarify your issues. If you have further trouble, feel free to provide more details, your input is always appreciated.

Best regards.

**Comment by r1cheu** [comment]
@glenn-jocher, thank you for your prompt reply. However, in my situation, the Yolov8 series works well on the same dataset. There are no issues with training, validation, or prediction. The classes used for training and validation are exactly the same.

**Comment by glenn-jocher** [comment]
@r1cheu thanks for the information! Is this issue reproducible on a common dataset like COCO8 or VOC? We can't reproduce issues on your custom dataset, and we are unable to debug issues we can not reproduce.

We've created a few short guidelines below to help users provide what we need in order to start investigating a possible problem.

## How to create a Minimal, Reproducible Example

When asking a question, people will be better able to provide help if you provide **code** that they can easily understand and use to **reproduce** the problem. This is referred to by community members as creating a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/). Your code that reproduces the problem should be:

- ✅ **Minimal** – Use as little code as possible to produce the problem
- ✅ **Complete** – Provide all parts someone else needs to reproduce the problem
- ✅ **Reproducible** – Test the code you're about to provide to make sure it reproduces the problem

For [Ultralytics](https://ultralytics.com/) to provide assistance your code should also be:

- ✅ **Current** – Verify that your code is up-to-date with GitHub with `git pull` or `pip install -U ultralytics` to ensure your problem has not already been solved in the latest code version.
- ✅ **Unmodified** – Your problem must be reproducible using official Ultrlaytics code without changes. [Ultralytics](https://ultralytics.com/) does not provide support for custom code ⚠️.

If you believe your problem meets all the above criteria, please close this issue and raise a new one using the 🐛 **Bug Report** [template](https://github.com/ultralytics/ultralytics/issues/new/choose) with a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us better understand and diagnose your problem. 

Thank you! 😃

**Comment by r1cheu** [comment]
@glenn-joche, thanks for your advice. I'll check it on COCO8.

**Comment by glenn-jocher** [comment]
@r1cheu, you're welcome! Running your process on COCO8 is a great idea. This dataset is widely used and reliable, making it an excellent benchmark for inspecting your model's behavior. If the same issue occurs when using COCO8, it may indicate a potential bug or inconsistency in the YOLOv8 implementation. If the issue doesn't manifest with COCO8, it could suggest that the problem is exclusive to your specific dataset or implementation. Either way, this step will provide valuable insight and guide next steps. Feel free to share the outcomes of your test, so we can assist you further. Happy coding!

**Comment by r1cheu** [comment]
Ah! It appears that when using the CLI `yolo val model=custom_rtdetr-l.pt`, it will simply use the ultralytics.YOLO instead of RTDETR to build the model. Everything works fine after using Python.

**Comment by glenn-jocher** [comment]
@r1cheu, I'm glad to hear that your issue was resolved by using Python instead of the CLI. Indeed, the `yolo val` CLI command is designed to use the default YOLO model for building and validation. When you want to use a custom model like RTDETR, you need to use Python so you can specify the precise model and parameters you want. Thank you for sharing this resolution with us and the wider community – these insights can be very valuable for others who might encounter the same issue. Happy coding!

**Comment by HaoLin97** [comment]
Hi @glenn-jocher , 
I am getting the same issue as @r1cheu. Whereby using the yolo val command through cli returns a 0 for all metrics and when I view the images from the validation, it shows as numbers instead of designated classes. I tried using CPU and using the python script but both resulted in the result of 0. I've attached screenshots below. I've also tried using the command yolo benchmark but that one just errored out - perhaps benchmark command does not support RT-DETR yet? What I am trying to do is to validate the rt-detr model that I retrained on 9 classes, the validation dataset is the same 9 classes. Any insight on what I should do to achieve this?

E.g. What the preds.jpg is showing
![image](https://github.com/ultralytics/ultralytics/assets/33756611/2109f684-41db-4abc-bee9-0ddb4331b94b)
vs what it should be(labels.jpg)
![image](https://github.com/ultralytics/ultralytics/assets/33756611/ee7a4848-f08f-4d45-8892-5bac17b6f40e)

**Comment by glenn-jocher** [comment]
Hi, 

It appears you are encountering a similar issue as @HaoLin97 encountered previously. When using the CLI `yolo val` command it was running Ultralytics YOLO by default instead of the RT-DETR model. This was overcome by running the process from within a Python script as it allows for more specificity in the model used.

The problem for you now is when running the validation with the same process but in Python you're still seeing zeros for all metrics. This suggests that there could be an issue with the validation data, possibly in the class indexing or naming. The pictures from validation being shown as numbers might suggest a mismatch between the real class names and those the model is predicting. 

For example, if your classes for training are [0, 1, 2] and the validation classes are [0, 1, 2, 3], it could result in an issue as no class '3' was present in the training set. This would also explain the KeyError during benchmarking since it's expecting classes that aren't present in the data. 

When retraining a model it's essential to ensure that the classes of your training and validation datasets match. Check your validation set for any misalignments in the classes used or the way they're indexed.

Also, regarding the benchmarking failure, you're correct in suggesting that the command does not support RT-DETR yet. The benchmark command is a relatively recent addition and may not be fully compatible with all models just yet. 

I hope this helps to clarify the situation. If anything is unclear or you have further questions please feel free to ask.

**Comment by HaoLin97** [comment]
Hi @glenn-jocher,

I have found the issue after seeing the issue from https://github.com/ultralytics/ultralytics/issues/4787. It seems like the issue is that yolo val requires the name of the model to contain rtdetr - after I changed the name of the model the validation works. I renamed the file from best.pt to redetr.pt. Although this may not be a bug, it should still be highlighted in the documentation.

All the best on the development and best wishes to the Ultralytics team.

**Comment by glenn-jocher** [comment]
@HaoLin97 hi,

I'm glad to hear you've discovered a solution! Indeed, your discovery about the requirement for the model name to contain 'rtdetr' is quite important and might be a valuable insight for other users as well. We appreciate you taking the time to bring this to our attention.

We apologize for any inconvenience this may have caused, and we'll definitely consider adding this point to our documentation for future reference. 

Your feedback is crucial in improving our models and documentation. Thank you for your kind words and best wishes.

Best regards,
The Ultralytics Team

**Comment by wangsun1996** [comment]
> @glenn-jocher, Ah! It appears that when using the CLI `yolo val model=custom_rtdetr-l.pt`, it will simply use the ultralytics.YOLO instead of RTDETR to build the model. Everything works fine after using Python.

hello，i have same question，could you please provide your code to solove this problem？

**Comment by r1cheu** [comment]
> > @glenn-jocher, Ah! It appears that when using the CLI `yolo val model=custom_rtdetr-l.pt`, it will simply use the ultralytics.YOLO instead of RTDETR to build the model. Everything works fine after using Python.
> 
> hello，i have same question，could you please provide your code to solove this problem？

Like I said, use Python, something like, 
```python
from ultralytics import RTDETR

model = RTDETR('/path/to/redetr.pt')  # load a custom model

metrics = model.val() 
```

**Comment by glenn-jocher** [comment]
@r1cheu, to resolve this issue, ensure you are using Python to specify your RTDETR model directly. The CLI may default to the standard YOLO model, which won't work for validation with a custom RTDETR. Using Python allows for a more tailored approach that recognizes your custom RTDETR model, leading to correct validation results. 

Please adapt and run similar Python code as suggested above with your model's correct path to perform validation. This should help you to validate your RTDETR model successfully.

**Comment by Hxiaowu666** [comment]
> 嗨，
> 
> 我在看到 #4787 的问题后发现了这个问题。似乎问题是 yolo val 要求模型的名称包含 rtdetr - 在我更改模型名称后，验证有效。我将文件从 best.pt 重命名为 redetr.pt。尽管这可能不是一个错误，但仍应在文档中突出显示。
> 
> 祝愿 Ultralytics 团队一切顺利，并祝愿 Ultralytics 团队一切顺利。

666

**Comment by glenn-jocher** [comment]
@Hxiaowu666 嗨！👋

感谢您分享您的发现和解决方案！确实，`yolo val` 对模型名称有特定要求是一个很好的提示，我们会考虑将这一点加入到我们的文档中，以帮助其他用户避免遇到相同的问题。感谢您的祝福，我们也祝您一切顺利！

祝编码愉快！🚀

-Ultralytics 团队

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in the linked PR (#20578). This update introduces several improvements to model handling in the Ultralytics codebase:

- The Model class now supports already-initialized Model instances, not just file paths or names.
- The YOLO class constructor is enhanced to handle Model objects directly, streamlining workflows.
- Automatic detection and conversion of RTDETR models within the YOLO class makes integration smoother and requires fewer manual steps.
- General improvements to flexibility and user-friendliness for working with custom or pre-initialized models.

Special thanks to the PR author and contributors for these enhancements!

To test whether these changes resolve your issue, you can upgrade to the latest version (8.3.141) with:
```
pip install -U ultralytics>=8.3.141
```
Alternatively, you can install the main branch directly:
```
pip install git+https://github.com/ultralytics/ultralytics.git@main
```
Feel free to wait for the next official PyPI release if preferred. Please let us know if this update resolves your issue or if you encounter any further problems. Thank you for reporting this and helping improve the project! 🙏
